ee/tabby-webserver/src/service/answer.rs

```diff
@@ -490,6 +512,85 @@ Remember, don't blindly repeat the contexts verbatim. When possible, give code s
     )
 }
 
+/// Combine code snippets from search results rather than utilizing multiple hits: Presently, there is only one rule: if the number of lines of code (LoC) is less than 200, and there are multiple hits (number of hits > 1), include the entire file.
+pub async fn merge_code_snippets(
+    repository: Option<Repository>,
+    hits: Vec<CodeSearchHit>,
+) -> Vec<CodeSearchHit> {
+    let Some(repository) = repository else {
+        return hits;
+    };
+
+    // group hits by filepath
+    let mut file_hits: HashMap<String, Vec<CodeSearchHit>> = HashMap::new();
+    for hit in hits.clone().into_iter() {
+        let key = format!("{}-{}", repository.source_id, hit.doc.filepath);
+        file_hits.entry(key).or_default().push(hit);
+    }
+
+    let mut result = Vec::with_capacity(file_hits.len());
+
+    for (_, file_hits) in file_hits {
+        if file_hits.len() > 1 {
+            // construct the full path to the file
+            let path: PathBuf = repository.dir.join(&file_hits[0].doc.filepath);
+            let file_content = match read_file_content(&path) {
+                Some(lines) => lines,
+                None => {
+                    //cannot read the file, just extend the hits
+                    result.extend(file_hits);
+                    continue;
+                }
+            };
+
+            if !file_content.is_empty() {
+                let mut insert_hit = file_hits[0].clone();
+                insert_hit.scores =
+                    file_hits
+                        .iter()
+                        .fold(CodeSearchScores::default(), |mut acc, hit| {
+                            acc.bm25 += hit.scores.bm25;
+                            acc.embedding += hit.scores.embedding;
+                            acc.rrf += hit.scores.rrf;
+                            acc
+                        });
+                // average the scores
+                let num_files = file_hits.len() as f32;
+                insert_hit.scores.bm25 /= num_files;
+                insert_hit.scores.embedding /= num_files;
+                insert_hit.scores.rrf /= num_files;
+                insert_hit.doc.body = file_content;
+                result.push(insert_hit);
+            }
+        } else {
+            result.extend(file_hits);
+        }
+    }
+    result
+}
+
+//Read file content and return raw file content string, it will return nothing if the file is over 200 lines
+pub fn read_file_content(path: &Path) -> Option<String> {
+    let file = match File::open(path) {
+        Ok(file) => file,
+        Err(e) => {
+            warn!("Error opening file {}: {}", path.display(), e);
+            return None;
+        }
+    };
+
+    let reader = BufReader::new(file);
+    let all_lines: Vec<String> = match reader.lines().collect::<Result<_, _>>() {
+        Ok(lines) => lines,
+        Err(_) => return None,
+    };
+
+    if all_lines.len() > 200 {
+        None
+    } else {
+        Some(all_lines.join("\n"))
```

Comment:
return raw content with BufReader, instead of join the content by lines (for example, the origin content might use CRLF ('\r\n'))